ecoli/processes/metabolism_redux.py

```diff
@@ -0,0 +1,499 @@
+        self.Sd = pd.DataFrame.from_dict(reactions, dtype=np.int8).fillna(0).astype(np.int8)
+        self.n_mets, self.n_orig_rxns = self.Sd.shape
+
+        # extract names of mets and rxns
+        self.mets = list(self.Sd.index)
+        self.rxns = list(self.Sd.columns)
+        self.intermediates = list(set(self.mets) - set(homeostatic_metabolites))
+
+        # steady state indices, secretion indices
+        self.intermediates_idx = [self.mets.index(met) for met in self.intermediates]
+
+
+    def set_up_exchanges(self,
+                         exchanges: Iterable[str],
+                         uptakes: Iterable[str]):
+
+        # explain what this does
+        all_exchanges = exchanges.copy()
+        all_exchanges.extend(uptakes)
+
+
+        self.Se = pd.DataFrame(index=self.Sd.index)
+        for met in all_exchanges:
+            exch_name = met + " exchange"
+            if met in uptakes:
+                self.Se[exch_name] = np.zeros(self.n_mets).astype(np.int8)
+                self.Se.at[met, exch_name] = 1
+
+            self.Se[exch_name + " rev"] = np.zeros(self.n_mets).astype(np.int8)
+            self.Se.at[met, exch_name + " rev"] = -1
+
+        self.S_orig = csr_matrix(np.array(self.Sd, dtype=np.float64))
+        self.S_exch = csr_matrix(np.array(self.Se, dtype=np.float64))
+
+        _, self.n_exch_rxns = self.S_exch.shape
+
+        self.secretion_idx = np.where(self.S_exch.sum(axis=0) == -1)[0]
+
+    def solve(self,
+              homeostatic_targets: Mapping[str, float],
+              maintenance_target: float,
+              kinetic_targets: Mapping[str, float],
+              binary_kinetic_targets: Mapping[str, float],
+              objective_weights: Mapping[str, float],
+              upper_flux_bound: float = 100,
+              **kwargs) -> FlowResult:
+
+        # objective targets: dict of dicts to floats, for now
+        ## binary kinetic
+        # starting fluxes: optional for now, could reintroduce
+        #
+
+        homeostatic_arr = [[self.mets.index(met), target] for met, target in homeostatic_targets.items()]
+        homeostatic_idx, homeostatic_target = np.array(homeostatic_arr, dtype=np.int64)[:, 0], np.array(homeostatic_arr)[:, 1]
+
+        kinetic_array = [[self.rxns.index(met), target] for met, target in kinetic_targets.items()]
+        kinetic_idx, kinetic_target = np.array(kinetic_array, dtype=np.int64)[:, 0], np.array(kinetic_array)[:, 1]
+
+        if binary_kinetic_targets:
+            binary_kinetic_idx = np.array([self.rxns.index(met) for met in binary_kinetic_targets.keys()], dtype=np.int64)
+        else :
+            binary_kinetic_idx = None
+
+        maintenance_idx = self.rxns.index("maintenance_reaction")  # TODO (Cyrus) - use name provided
+
+        # set up variables
+        v = cp.Variable(self.n_orig_rxns)
+        e = cp.Variable(self.n_exch_rxns)
+        dm = self.S_orig @ v + self.S_exch @ e
+        exch = self.S_exch @ e
```

Comment:
I see what you meant now by the easy-to-understand equation syntax. This, the constraints, and the loss function code are really nice.

Reply:
Yep!